### PR TITLE
TEL-5956: Build SIPREC SRC Module

### PR DESCRIPTION
**Pull request by baloeng**
This is the first draft of the siprec src module. Kindly review and share feedback

### CODE REVIEW

File: src/mod/applications/mod_siprec_src/mod_siprec_src.c [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+	const char *remote_media_ip = NULL;
+	switch_port_t remote_rtp_port_r = 0, remote_rtp_port_w = 0;
+	char *originate_str = NULL;
+	const char *sdp_in;
+	uint32_t timeout = 60;
+	int rc = 0;
+
+	channel = switch_core_session_get_channel(session);
+	SWITCH_STANDARD_STREAM(sdp);
+	SWITCH_STANDARD_STREAM(rs);
```
[the remark this answers is not in the document]
Comment (baloeng) [reply src/mod/applications/mod_siprec_src/mod_siprec_src.c:348]: Yeah, I fixed this in the last commit

File: src/mod/applications/mod_siprec_src/mod_siprec_src.c [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+		int i = 0;
+
+		ptr += 7;
+		// remove m=audio to catch the next one
+		*r_sdp = strdup(ptr);
+
+		while(*ptr == ' ') {
+			ptr++;
+		}
+		while(*ptr && *ptr != ' ' && *ptr != '\r' && *ptr != '\n') {
```
[the remark this answers is not in the document]
Comment (baloeng) [reply src/mod/applications/mod_siprec_src/mod_siprec_src.c:143]: Good catch. I'll add a limit to the loop

File: src/mod/endpoints/mod_sofia/sofia_media.c [hunk 3]
```diff
@@ -122,8 +122,14 @@ static void process_mp(switch_core_session_t *session, switch_stream_handle_t *s
...
-		if (drop_sdp && (strcasecmp(dname, "application/sdp") == 0)) {
-			return;
+		
+		//if (drop_sdp && (strcasecmp(dname, "application/sdp") == 0)) {
+			//return;
+		//}
+
+		// Allow custom SDP
+		if (strcasecmp(dname, "application/sdp") == 0) {
+			allow_sdp = SWITCH_TRUE;
```
[the remark this answers is not in the document]
Comment (baloeng) [reply src/mod/endpoints/mod_sofia/sofia_media.c]: Done

File: src/mod/endpoints/mod_sofia/sofia_media.c [hunk 4]
```diff
@@ -115,14 +113,14 @@ SWITCH_DECLARE_NONSTD(switch_status_t) switch_string_stream_write(switch_stream_
...
-static void process_mp(switch_core_session_t *session, switch_stream_handle_t *stream, const char *boundary, const char *str, const char*isup, intptr_t isup_len, switch_bool_t drop_sdp) {
+static void process_mp(switch_core_session_t *session, switch_stream_handle_t *stream, const char *boundary, const char *str, const char*isup, intptr_t isup_len, switch_bool_t drop_sdp, switch_bool_t allow_sdp) {
 	char *dname = switch_core_session_strdup(session, str);
 	char *dval;
 
 	if ((dval = strchr(dname, ':'))) {
 		*dval++ = '\0';
 		
-		if (drop_sdp && (strcasecmp(dname, "application/sdp") == 0)) {
+		if (drop_sdp && (strcasecmp(dname, "application/sdp") == 0) && !allow_sdp) {
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_media.c:123]: drop_sdp and not allow_sdp seems to be similar purpose. Why not just use drop_sdp instead. See my other comment for details

File: src/mod/endpoints/mod_sofia/sofia_media.c [hunk 5]
```diff
@@ -162,11 +163,11 @@ char *sofia_media_get_multipart(switch_core_session_t *session, const char *pref
 					int i = 0;
 
 					for(i = 0; i < hi->idx; i++) {
-						process_mp(session, &stream, boundary, hi->array[i], isup, isup_len, !!sdp);
+						process_mp(session, &stream, boundary, hi->array[i], isup, isup_len, !!sdp, allow_sdp);
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_media.c:166]: We can reuse the function parameter drop_sdp having this variables before calling this method
```
switch_bool_t allow_custom_sdp = switch_true(switch_channel_get_variable(channel, "allow_custom_sdp"));
switch_bool_t drop_sdp = !!sdp && !allow_custom_sdp;
...
process_mp(session, &stream, boundary, value, isup, isup_len, drop_sdp);
```

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/mod/endpoints/mod_sofia/sofia_media.c [hunk 6]
```diff
@@ -147,10 +145,13 @@ char *sofia_media_get_multipart(switch_core_session_t *session, const char *pref
 	int x = 0;
 	switch_channel_t *channel = switch_core_session_get_channel(session);
 	const char *boundary = switch_core_session_get_uuid(session);
+	switch_bool_t allow_sdp;
 
 	char * isup = (char*)switch_channel_get_private(channel, "_isup_payload");
 	intptr_t isup_len = (intptr_t)switch_channel_get_private(channel, "_isup_payload_size");
-	
+	allow_sdp = switch_true(switch_channel_get_variable(channel, "siprec_custom_sdp"));
```
[the remark this answers is not in the document]
Comment (baloeng) [reply src/mod/endpoints/mod_sofia/sofia_media.c]: Fixed

File: src/mod/endpoints/mod_sofia/sofia_media.c (hunk 5 above)
Comment (baloeng) [reply src/mod/endpoints/mod_sofia/sofia_media.c]: Done

**Review by minhtuan1407-telnyx — APPROVED** [review]

**Review by dev-ryanc — APPROVED** [review]

**Pull request merged**